libs/contrib/Text/Parser.idr

```diff
@@ -108,6 +109,16 @@ mutual
   count (Qty (S min) (Just Z)) _ = fail "Quantity out of order"
   count (Qty (S min) (Just (S max))) p = count1 (between min max) p
 
+||| Parse `p` `n` times, returning the vector of values.
+export
+countExactly : (n : Nat) ->
+               (p : Grammar tok True a) ->
+               Grammar tok (isSucc n) (Vect n a)
+countExactly Z p = Empty []
+countExactly (S k) p = do x <- p
+                          seq (countExactly k p)
+                              (\ xs => pure (x :: xs))
```

Comment:
```suggestion
                          xs <- countExactly k p
                          pure (x :: xs)
```

Reply:
Why not `countExactly (S k) p = [| p :: (countExactly k p) |]` then?

Reply:
This did seem like a good idea until I tried it:

```idris
countExactly : (n : Nat) ->
               (p : Grammar tok True a) ->
               Grammar tok (isSucc n) (Vect n a)
countExactly Z p = Empty []
countExactly (S k) p = do x <- p
                          xs <- countExactly k p
                          pure ?xConsXs

-- results in this compiler message
Text/Parser.idr:119:27--119:49:While processing right hand side of countExactly at Text/Parser.idr:118:1--127:1:
Can't solve constraint between:
	(xs : ?_) -> Grammar ?tok False ?ty
and
	inf (isSucc k) (Vect k a -> Grammar tok ?c2 (Vect (S k) a))
at:
119	                          xs <- countExactly k p
	                          ^^^^^^^^^^^^^^^^^^^^^^
```

I'm not exactly sure why it doesn't work, but I think it might have to do with `(isSucc k)`, because this compiles fine:

```idris
countExactly : (n : Nat) ->
               (p : Grammar tok True a) ->
               Grammar tok (isSucc n) (Vect n a)
countExactly Z p = Empty []
countExactly (S Z) p = do x <- p
                          pure [x]
countExactly (S (S k)) p = do x <- p
                              xs <- countExactly (S k) p
                              pure $ x :: xs
```

Anyways I think it's more clear what is going on when keeping `seq`.

Reply:
I'm sorry to intervene once more, but my simplification of Guillaume's suggestion that uses idiom brackets typechecks without `S (S k)` stuff.

Reply:
> I'm sorry

Don't be, I was not familiar with the `[| |]` syntax and assumed it would be equivalent to the do-notation.


Actually this results in something slightly different, as `[| |]` uses `<*>`, instead of `>>=`.
Do-notation desugars to `>>=` which uses `SeqEat` for *consuming grammars* and `[| |]` desugars to `<*>` which uses `SeqEmpty x (\f => map f (Force y))`.

This shouldn't matter at all though I think :)

I'll change it.